packages/blocks/src/surface-block/elements/group/group-element.ts

```diff
@@ -0,0 +1,217 @@
+import type { IVec } from '../../utils/vec.js';
+import type { HitTestOptions } from '../edgeless-element.js';
+import { SurfaceElement } from '../surface-element.js';
+import {
+  getFontString,
+  getLineHeight,
+  getLineWidth,
+  truncateTextByWidth,
+} from '../text/utils.js';
+import type { IGroup, IGroupLocalRecord } from './types.js';
+
+export class GroupElement extends SurfaceElement<IGroup, IGroupLocalRecord> {
+  private _cacheChildren: string[] = [];
+  private _cacheId: string = '';
+  private _titleHeight = getLineHeight("'Kalam', cursive", 16);
+  private _titleWidth = 0;
+  private _padding = [0, 0];
+  private _radius = 0;
+
+  override containedByBounds(bound: Bound): boolean {
+    return bound.contains(Bound.deserialize(this.xywh));
+  }
+
+  override intersectWithLine(start: IVec, end: IVec): PointLocation[] | null {
+    const bound = Bound.deserialize(this.xywh);
+    return linePolygonIntersects(start, end, bound.points);
+  }
+
+  override init(): void {
+    super.init();
+    const { surface } = this;
+    this._cacheId = this.id;
+    this.childrenElements.forEach(ele => {
+      surface.setGroup(ele, this.id);
+    });
+    this._cacheChildren = this.children;
+    const children = this.yMap.get('children') as IGroup['children'];
+    children.observe(event => {
+      for (const [key, { action }] of Array.from(event.changes.keys)) {
+        if (action === 'delete') {
+          const child = this.surface.pickById(key);
+          if (child && surface.getGroup(child) === this.id)
+            surface.setGroup(child, surface.getGroup(this));
+        } else if (action === 'add') {
+          const child = this.surface.pickById(key);
+          assertExists(child);
+          surface.setGroup(child, this.id);
+        } else {
+          console.log('unexpected', key);
+        }
+      }
+    });
+  }
+
+  override get xywh() {
+    const { surface } = this;
+    const children = this.children;
+    if (children.length === 0) return '[0,0,0,0]';
+    const bound: Bound = children.reduce((prev, cur) => {
+      const ele = surface.pickById(cur);
+      assertExists(ele);
+      return prev.unite(ele.gridBound);
+    }, surface.pickById(children[0])!.gridBound);
+    return bound.serialize();
+  }
+
+  get title() {
+    return this.yMap.get('title') as IGroup['title'];
+  }
+
+  get children() {
+    return Array.from((<IGroup['children']>this.yMap.get('children')).keys());
+  }
+
+  get childrenElements() {
+    return this.children.map(id => this.surface.pickById(id)!);
+  }
+
+  get titleHeight() {
+    return this._titleHeight;
+  }
+
+  get titleWidth() {
+    return this._titleWidth;
+  }
+
+  get padding() {
+    return this._padding;
+  }
+
+  get radius() {
+    return this._radius;
+  }
+
+  override get gridBound() {
+    const bound = Bound.deserialize(this.xywh);
+    bound.y -= this._titleHeight;
+    bound.h += this._titleHeight;
+    return bound;
+  }
+
+  addChild(id: string) {
```

Comment:
Why put the method on the group element instance side, instead of adding method in `GroupManager`?

Reply:
1. the interface is simple.
2. groupElement updated its internal cache state.